### PR TITLE
Introduced adversarial page cache

### PR DESCRIPTION
**Pull request by lutovich**
Such page cache implementation is able to wrap the given page cache and introduce some misbehaviour to exercise the user code. It can either throw `RuntimeException`s/`IOException`s or return a page cursor with inconsistent reads. Misbehaviour is controlled by the given `Adversary`.

Changed `PageCacheRule` to use the `AdversarialPageCache`.

### CODE REVIEW

**Comment by lutovich** [comment]
@chrisvest could you please take a look at this PR?

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    public AdversarialPageCache( PageCache delegate, Adversary adversary )
+    {
+        this.delegate = Objects.requireNonNull( delegate );
+        this.adversary = Objects.requireNonNull( adversary );
+    }
+
+    @Override
+    public PagedFile map( File file, int pageSize, OpenOption... openOptions ) throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java]: You shouldn't throw `FileNotFoundException` if the `StandardOpenOption.CREATE` is present in the `openOptions`.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        delegate.flushAndForce();
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
+        delegate.close();
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java:80]: I also don't think `flushAndForce` and `close` can throw `FileNotFoundException`, or `SecurityException` for that matter. We should only throw exceptions that can actually happen.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        adversary.injectFailure( SecurityException.class );
+        return delegate.pageSize();
+    }
+
+    @Override
+    public int maxCachedPages()
+    {
+        adversary.injectFailure( SecurityException.class );
+        return delegate.maxCachedPages();
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java]: `pageSize` and `maxCachedPages` have no conceivable path that leads to exceptions being thrown.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        adversary.injectFailure( SecurityException.class );
+        delegate.putShort( value );
+    }
+
+    @Override
+    public void putShort( int offset, short value )
+    {
+        adversary.injectFailure( SecurityException.class );
+        delegate.putShort( offset, value );
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: The `get*` and `put*` methods can only throw `IndexOutOfBoundsException`.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+    @Override
+    public void setOffset( int offset )
+    {
+        adversary.injectFailure( SecurityException.class );
+        delegate.setOffset( offset );
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: `setOffset` can only throw `IndexOutOfBoundsException`.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        adversary.injectFailure( SecurityException.class );
+        return delegate.getCurrentPageSize();
+    }
+
+    @Override
+    public File getCurrentFile()
+    {
+        adversary.injectFailure( SecurityException.class );
+        return delegate.getCurrentFile();
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: `getOffset`, `getCurrentPageId`, `getCurrentPageSize` and `getCurrentFile` don't throw anything.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+    @Override
+    public void rewind() throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
+        delegate.rewind();
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: `rewind` don't throw anything.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        return delegate.next();
+    }
+
+    @Override
+    public boolean next( long pageId ) throws IOException
+    {
+        currentReadIsInconsistent = adversary.injectFailureOrMischief( FileNotFoundException.class, IOException.class,
+                SecurityException.class );
+        return delegate.next( pageId );
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: The `next` methods can also throw `IllegalStateException`, in addition to what is currently listed.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+    @Override
+    public void close()
+    {
+        adversary.injectFailure( SecurityException.class );
+        delegate.close();
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: I don't think `close` can throw anything.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+    @Override
+    public boolean shouldRetry() throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCursor.java]: Like the `next` methods, `shouldRetry` can also throw `IllegalStateException`.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        this.adversary = Objects.requireNonNull( adversary );
+    }
+
+    @Override
+    public PageCursor io( long pageId, int pf_flags ) throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
+        PageCursor pageCursor = delegate.io( pageId, pf_flags );
+        return new AdversarialPageCursor( pageCursor, adversary );
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java]: `io` can only throw `IllegalStateException`

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public int pageSize()
+    {
+        adversary.injectFailure( SecurityException.class );
+        return delegate.pageSize();
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java]: `pageSize` can't throw.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java (hunk 2 above)
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPageCache.java:80]: Wait, I guess `flushAndForce` can throw `FileNotFoundException` and `SecurityException` after all.

File: community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
+        delegate.flushAndForce();
+    }
+
+    @Override
+    public long getLastPageId() throws IOException
+    {
+        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class );
+        return delegate.getLastPageId();
+    }
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialPagedFile.java]: `getLastPageId` can only throw `IllegalStateException`.

**Comment by chrisvest** [comment]
@lutovich You also have some test failures.

**Comment by lutovich** [comment]
@chrisvest mentioned issues should be fixed now. I've also made 2 implementations of the `PageCursor` - `AdversarialReadPageCursor` and `AdversarialWritePageCursor`. First one can't write and can produce inconsistent reads and throw. Second one always produces consistent reads and can throw.

**Comment by chrisvest** [comment]
@lutovich Perfect!

**Pull request merged**